### PR TITLE
Search index tools KB article

### PR DESCRIPTION
**Pull request by erincdustin**
KB article for walking through new search index tools feature in Portal

### CODE REVIEW

File: content/documents/search-index-tools.mdx [hunk 1]
```diff
@@ -0,0 +1,73 @@
...
+title:          Introducing Search Index Tools for Portal
+description:    A new set of tools for troubleshooting search index issues
+author:         edustin 
+publishDate:    2023-01-03
+updatedDate:    2023-01-03
+tags:           ["Search", "Portal"]
+---
+
+We're excited to unveil a long-requested enhancement for OrderCloud's premium search feature: a self-service tool for viewing and rebuilding your search indexes. By the end of this article, you should be ready to investigate and resolve basic search errors in your marketplace without the need for OrderCloud support.
+
```
Comment (mirandaposthumus) [inline content/documents/search-index-tools.mdx:12]: comma after marketplace and then "eliminating the need to seek additional support from Sitecore in most cases."

File: content/documents/search-index-tools.mdx [hunk 2]
```diff
@@ -0,0 +1,73 @@
...
+type:           article
+title:          Introducing Search Index Tools for Portal
+description:    A new set of tools for troubleshooting search index issues
+author:         edustin 
+publishDate:    2023-01-03
+updatedDate:    2023-01-03
+tags:           ["Search", "Portal"]
+---
+
+We're excited to unveil a long-requested enhancement for OrderCloud's premium search feature: a self-service tool for viewing and rebuilding your search indexes. By the end of this article, you should be ready to investigate and resolve basic search errors in your marketplace without the need for OrderCloud support.
```
Comment (mirandaposthumus) [inline content/documents/search-index-tools.mdx:11]: change "a long-requested enhancement" to "an exciting enhancement"

File: content/documents/search-index-tools.mdx [hunk 3]
```diff
@@ -0,0 +1,73 @@
...
+
+## A Background On Search Indexes
+
+The following section builds upon [previous articles](/knowledge-base?t=Search) for OrderCloud's premium search capabilities. Understanding these concepts will help you interpret your marketplace's search indexes.
+
```
Comment (mirandaposthumus) [inline content/documents/search-index-tools.mdx:16]: The following section builds upon [previous content](/knowledge-base?t=Search) on OrderCloud's premium search capabilities. Understanding these concepts will be helpful when troubleshooting your marketplace's search indexes.

File: content/documents/search-index-tools.mdx [hunk 4]
```diff
@@ -0,0 +1,73 @@
...
+- Premium search endpoints serve cached Product and Order data from Elasticsearch to support greater search capabilities and faster performance.
```
Comment (mirandaposthumus) [inline content/documents/search-index-tools.mdx:17]: submitted Order data

File: content/documents/search-index-tools.mdx [hunk 5]
```diff
@@ -0,0 +1,73 @@
...
+- Only one "active" search index per type can exist at any given time, and only an index with a status of `Ready` can be the active index for a marketplace. OrderCloud will point to a new index only after a successful rebuild.
```
Comment (mirandaposthumus) [inline content/documents/search-index-tools.mdx:18]: per type (Product or Order)

File: content/documents/search-index-tools.mdx [hunk 6]
```diff
@@ -0,0 +1,73 @@
...
+- Index errors typically arise when xp data types are not consistent between items within the same resource type. See the section on [xp schema errors](/knowledge-base/search-index-tools#understanding-xp-schema-errors) below for more detail.
+- Errors can exist on an active index. OrderCloud will fail an index build if the number of rejected items is considered too high or xp limits are exceeded.
```
Comment (mirandaposthumus) [inline content/documents/search-index-tools.mdx:20]: what does "xp limits exceeded" mean?

File: content/documents/search-index-tools.mdx [hunk 7]
```diff
@@ -0,0 +1,73 @@
...
+- A full index rebuild is a tool to "reset" your search index from the master OrderCloud record, and can oftentimes resolve data synchronization issues. This capability was previously limited to only the support team, but is now extended to marketplace administrators [with limitations](/knowledge-base/search-index-tools#rebuilding-your-index).
```
Comment (mirandaposthumus) [inline content/documents/search-index-tools.mdx:21]: what are "data synchronization issues"?

File: content/documents/search-index-tools.mdx [hunk 8]
```diff
@@ -0,0 +1,73 @@
...
+
+## Troubleshooting With Search Index Tools
+
+Registered portal users now have visibility to search index details for their assigned marketplaces. After selecting a marketplace from the list, you will have the option to view either Product or Order search indexes from the tab selections.
```
Comment (mirandaposthumus) [inline content/documents/search-index-tools.mdx:25]: above you say this is limited to Marketplace administrators, here you say registered portal users + assigned marketplaces. Let's be consistent/clear on who has access to these features.

File: content/documents/search-index-tools.mdx [hunk 9]
```diff
@@ -0,0 +1,73 @@
...
+
+## Troubleshooting With Search Index Tools
+
+Registered portal users now have visibility to search index details for their assigned marketplaces. After selecting a marketplace from the list, you will have the option to view either Product or Order search indexes from the tab selections. 
+
+<figure class="blog-figure">
+<img src="/images/knowledge-base/inline/search-index-tools/product_index_view.png" alt="Search Index Tools Products View"/>
+</figure>
+
+The details view for a selected search index will display the following:
```
Comment (mirandaposthumus) [inline content/documents/search-index-tools.mdx:31]: detail

File: content/documents/search-index-tools.mdx [hunk 10]
```diff
@@ -0,0 +1,73 @@
...
+- Status - `Ready`, `Building`, `Rebuilding`, `WaitingToBuild`, `Failed`, `Canceled`, or `Invalid`
+- The `Active` badge - indicates the "current" index OrderCloud is pointed to for the selected resource
+- Schema errors (if any) associated with the index
+  - Visibility to error logs allows users to target which items are causing issues for a given index. 
+  - Logs may include the rejected Product, Order, or Line Item ID, the flagged xp field, and its corresponding value.
+  - Error logs are limited to 100 documents for brevity, although there may be more items in the marketplace that have failed to index in Elasticsearch.
+
+>Not seeing a search index for your resource type? You will need to create items for that resource first (reminder that only submitted Orders are indexed) and make a LIST endpoint call. Also keep in mind that premium search for Orders [may not be enabled](/knowledge-base/order-search#how-do-i-test-this) for your Production marketplace if it was created prior to April 20, 2022.
```
Comment (mirandaposthumus) [inline content/documents/search-index-tools.mdx:39]: don't want to capitalize `LIST` as it's not an http verb - "make a call to list either products or orders depending on the index type".

File: content/documents/search-index-tools.mdx [hunk 11]
```diff
@@ -0,0 +1,73 @@
...
+
+## Understanding Xp Schema Errors
+
+Let's walk through an example of an xp schema inconsistency. Two products are created, both with the same xp field name, but with different datatypes. This inconsistency produces the following error: _"Failed to parse field [xp.FreeShipping] of type [boolean] in document with id 'product2'. Preview of field's value: '0'"_
```
Comment (mirandaposthumus) [inline content/documents/search-index-tools.mdx:43]: we always want to talk about xp as key/value pairs, so be sure to use that language throughout 

xp key name, but with different value types (e.g. an integer and a boolean).

File: content/documents/search-index-tools.mdx [hunk 12]
```diff
@@ -0,0 +1,73 @@
...
+
+`product1.xp.FreeShipping`: `false`  
+`product2.xp.FreeShipping`: `0`
+
+Elasticsearch dynamically maps xp datatypes at index time and will reject incoming documents when it cannot [coerce](https://www.elastic.co/guide/en/elasticsearch/reference/current/coerce.html) the xp value to the previously determined types. In this example `product1` was created and indexed first. As a result, Elasticsearch was not able to coerce the value of `product2.xp.FreeShipping` (type `int`) to a `boolean`, and it was excluded from the index due to an inconsistent datatype.
```
Comment (mirandaposthumus) [inline content/documents/search-index-tools.mdx:48]: I don't know that we want to link to Elasticsearch documentation here, it's an implementation concern that isn't really important for users to know about, just understand how/why it works a given way. Additionally, for submitted orders, this is the opposite. xp value types are "last in" (where as products are "first in").

File: content/documents/search-index-tools.mdx [hunk 13]
```diff
@@ -0,0 +1,73 @@
...
+
+If you encounter an index error in your marketplace, you will notice that search results do not behave as expected. In this example, `GET v1/products/product2` would return the product as expected, but that same product would not appear in a `GET v1/products` list.
```
Comment (mirandaposthumus) [inline content/documents/search-index-tools.mdx:50]: change "do not" to "may not"

**Review by mirandaposthumus — CHANGES_REQUESTED** [review]

File: content/documents/search-index-tools.mdx (hunk 8 above)
Comment (erincdustin) [reply content/documents/search-index-tools.mdx:25]: Any registered portal user can view search index details of marketplaces that they are a member of.  only administrators can trigger a rebuild.  I'll look at the language to make this clearer

Comment (mirandaposthumus) [reply content/documents/search-index-tools.mdx:25]: I'm not sure I agree with that choice - I feel like you should have to have a specific role to even see them.

**Review by mirandaposthumus — APPROVED** [review]

**Pull request merged**